http-verbs-common/src/main/scala/uk/gov/hmrc/http/client2/package.scala

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.http
+
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+
+package client2 {
+  trait Streaming
+}
+package object client2
+  extends StreamHttpReadsInstances {
+
+  // ensures strict HttpReads are not passed to stream function, which would lead to stream being read into memory
+  // (or runtime exceptions since HttpResponse.body with throw exception for streamed responses)
+  type StreamHttpReads[A] = HttpReads[A] with Streaming
+}
+
+trait StreamHttpReadsInstances {
+  def tag[A](instance: A): A with client2.Streaming =
+    instance.asInstanceOf[A with client2.Streaming]
+
+  implicit val readEitherSource: HttpReads[Either[UpstreamErrorResponse, Source[ByteString, _]]] with client2.Streaming =
```

Comment:
It looks like these instances are defined in terms of `HttpResponse.bodyAsSource`, which looks like it transforms a strict `String`:

```scala
  def bodyAsSource: Source[ByteString, _] =
    Source.single(ByteString(body))
```

Reply:
Ignore this, I just spotted the override, sorry!

Reply:
The default `bodyAsSource` is defined against the strict payload, however if the `HttpResponse` was created with a `Source`, it will return that instead.
When calling `client.execute`, play-ws will provide a strict payload - if you do call `bodyAsSource` it will just convert the strict payload into a Stream.
When calling `client.stream`, the `HttpResponse` will be created with the Stream - if you call `body` it will throw an exception rather than load the stream all into memory - although, this should be prevented by being coerced into using `StreamHttpReads` rather than just any `HttpReads`